samza-core/src/main/scala/org/apache/samza/coordinator/JobModelManager.scala

```diff
@@ -19,98 +19,183 @@
-    currentJobModelManager
+    * Builds the {@see GrouperContext} for the samza job.
+    * @param config represents the configurations defined by the user.
+    * @param localityManager provides the processor to host mapping persisted to the metadata store.
+    * @param taskAssignmentManager provides the processor to task assignments persisted to the metadata store.
+    * @return the instantiated {@see GrouperContext}.
+    */
+  private def getGrouperContext(config: Config, localityManager: LocalityManager, taskAssignmentManager: TaskAssignmentManager) = {
+    val processorLocality: util.Map[String, LocationId] = getProcessorLocality(config, localityManager)
+    val taskAssignment: util.Map[String, String] = taskAssignmentManager.readTaskAssignment()
+    val taskNameToProcessorId: util.Map[TaskName, String] = new util.HashMap[TaskName, String]()
+    for ((taskName, processorId) <- taskAssignment) {
+      taskNameToProcessorId.put(new TaskName(taskName), processorId)
+    }
+
+    val taskLocality:util.Map[TaskName, LocationId] = new util.HashMap[TaskName, LocationId]()
+    for ((taskName, processorId) <- taskAssignment) {
+      if (processorLocality.containsKey(processorId)) {
+        taskLocality.put(new TaskName(taskName), processorLocality.get(processorId))
+      }
+    }
+    val grouperContext = new GrouperContext(processorLocality, taskLocality, new util.HashMap[TaskName, util.List[SystemStreamPartition]](), taskNameToProcessorId)
+    grouperContext
   }
 
   /**
-   * For each input stream specified in config, exactly determine its
-   * partitions, returning a set of SystemStreamPartitions containing them all.
-   */
-  private def getInputStreamPartitions(config: Config, streamMetadataCache: StreamMetadataCache) = {
+    * Retrieves and returns the processor locality of a samza job using provided {@see Config} and {@see LocalityManager}.
+    * @param config provides the configurations defined by the user. Required to connect to the storage layer.
+    * @param localityManager provides the processor to host mapping persisted to the metadata store.
+    * @return the processor locality.
+    */
+  private def getProcessorLocality(config: Config, localityManager: LocalityManager) = {
+    val containerToLocationId: util.Map[String, LocationId] = new util.HashMap[String, LocationId]()
+    val existingContainerLocality = localityManager.readContainerLocality()
+
+    for (containerId <- 0 to config.getContainerCount) {
+      val localityMapping = existingContainerLocality.get(containerId.toString)
+      // To handle the case when the container count is increased between two different runs of a samza-yarn job,
+      // set the locality of newly added containers to any_host.
+      var locationId: LocationId = new LocationId("ANY_HOST")
+      if (localityMapping != null) {
+        locationId = new LocationId(localityMapping.get(SetContainerHostMapping.HOST_KEY))
```

Comment:
Is it guaranteed that `localityMapping.get(SetContainerHostMapping.HOST_KEY)` returns non-null value? Otherwise we would be throwing an IllegalArgumentException instead of using `ANY_HOST` as the location id. Do we want to fall back to `ANY_HOST`?